openedx/core/djangoapps/discussions/docs/decisions/0002-organization-wide-provider-filtering.rst

```diff
@@ -0,0 +1,155 @@
+Allow organization-wide filtering of discussion providers
+=========================================================
+
+
+Status
+------
+
+Proposal
+
+
+Context
+-------
+
+As part of the BD-03 initiative (Blended Development, Project 3),
+we want to allow Organizations to filter the list of available
+third-party discussion providers.
+
+This could be due to a number of motivations, eg:
+- desired experience
+- support levels
+- data sharing
+
+We would, however, want to allow for manual exceptions.
+
+This document proposes a new configuration system that allows
+admins to maintain a deny-allow-list to allow filtering of discussions
+providers on a per-organization basis.
+
+
+Requirements
+------------
+
+Given a particular organization, we must be able to remove options
+from the list of available discussions providers, during configuration.
+
+
+Consideration
+-------------
+
+- Django settings based (remote config, etc.)
+  - pros:
+    - ease of implmentation
+  - cons:
+    - overhead in maintenance (even with remote config, it would require
+      an engineer to make a code/config change and pull request
+- Database-backed
+  - pros:
+    - ease of use
+      - by (django) admins
+  - cons:
+    - possible schema lock-in
+      - eg: do we scope it to `organizations` only?
+            or do we generalize it enough to use elsewhere?
+
+
+Decision
+--------
+
+We propose to implement this as a Django database-backed model,
+`ProviderFilter`.
+
+The underlying model can be represented as:
+
+.. code-block:: python
+    class ProviderFilter(StackedConfigurationModel):
+        allow = models.CharField(blank=True, ...)
+        deny = models.CharField(blank=True, ...)
+
+- `allow` is a space-delimited string of providers to allow
+
+- `deny` is a space-delimited string of providers to deny
+
+By extending the `StackedConfigurationModel`, we gain the flexibility of
+stacking global/site/org/course/run values to not only provide
+organization-wide filtering today, but to arbitrary levels of
+granularity in the future. As this base model already exists in
+`edx-platform`, integration is low-effort.
+
+
+Examples
+--------
+
+With this system, we can configure global/site-wide defaults,
+
+.. code-block:: python
+    ProviderFilter.objects.create()
+
+And then override to deny an external provider for an organization:
+
+.. code-block:: python
+    ProviderFilter.objects.create(
+        org='Piazza-Less',
+        deny='lti-providerx',
+    )
+
+Or override to deny _all_ external providers for an organization:
+
+.. code-block:: python
+    ProviderFilter.objects.create(
+        org='InternalOrganization',
+        allow='cs_comments_service',
+    )
+
+And grant an exemption to a specific course:
+
+.. code-block:: python
+    ProviderFilter.objects.create(
+        # a course in 'InternalOrganization'
+        course=course,
+        allow='cs_comments_service lti-providerx',
+    )
+
+
+Logic
+-----
+
+.. code-block:: python
```

Comment:
Sanity check, I read this code block as saying that the deny-list entries supercede the allow-list entries, not vice versa, since they get the last crack at the `available` list.  That sound right?  I wonder if there's a way to set this up without that ambiguity... though it's low risk for an admin to get it wrong, presumably, since it just means something will show up/not show up.

Reply:
I had the same confusion as David and your explanation makes sense; I'll take a look at the Django admin help text to see if we can clarify. I think what we want is basically - "By default, every option is allowed. Creating an `allow_list` will limit available options to only those present in the `allow_list`. Creating a `deny_list` will explicitly exclude all options in the `deny_list`". Is my understanding correct here?

Reply:
Actually @stvstnfrd get back to me on the above, and I'll assist with the help text.